### DISCUSSION TITLE
x-modelable inside of a tree created with x-html reloads the entire form

### DISCUSSION THREAD

**Discussion by tobia** [alpinejs/alpine#3160] [7. Bugs (archived)]
Hello

I often use `x-html` to generate and mount parts of an application at runtime. For example, when a user opens a panel of some sort, I may request the panel's HTML source code from a PHP page using `fetch()` and mount it inside the panel with `x-html`. This is more lightweight than using an iframe and allows richer interaction between the various parts of the application.

At the same time, some parts of the application that are loaded this way may contain forms with custom components that are glued together using `x-modelable`.

This combination worked fine up to Alpine 3.9.2, but became broken from Alpine 3.9.3 onwards (still broken in 3.10.3)

Here is a minimal HTML page to reproduce the bug. In this page I'm reading the inner HTML from a string constant instead of loading it from PHP, but the results are the same:

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <title>Test bug Alpine.js</title>
    <!-- 3.9.2 WORKS, 3.9.3 FAILS -->
    <script defer src="https://unpkg.com/alpinejs@3.9.3/dist/cdn.min.js"></script>
  </head>
  <body>
    <script>
      THE_HTML = `
        <div x-data="{ text: '', counter: 0 }">
          <p>
            First write something here:
            <input x-model="text">
          </p>
          <p>
            Then click this:
            <button x-data="{ n: 0 }" @click="n++" x-modelable="n" x-model="counter">
              CLICK ME
            </button>
            <span x-text="counter"></span>
          </p>
          <p>
            If the text disappears, you have the bug.
          </p>
        </div>
      `
    </script>

    <h3>Bug showcase:</h3>
    <div x-data="{ html: THE_HTML }" x-html="html"></div>

    <hr>

    <h3>Workaround:</h3>
    <div
      x-data="{ html: null }"
      x-init="$watch('html', html => Alpine.effect(() => { $root.innerHTML = html }))"
    >
      <button @click="html = THE_HTML">START</button>
    </div>
  </body>
</html>
```

I have identified this change to be the culprit: https://github.com/alpinejs/alpine/commit/3fe20869e1671171fdf7b32787dd4e97e1f1ed55

Unfortunately I tried to understand what `effect()`, `mutateDom()`, and `initTree()` are supposed to do in Alpine, and why `x-modelable` in particular breaks the entire thing, but it's way over my head!

After finding the source of the bug, I switched to the workaround shown inside the example: changing `$root.innerHTML` directly in a `$watch` instead of using `x-html`. I added `Alpine.effect()` because it was there in the original code, but I have no idea what it's for.

This took me so much time to figure out that I think it should be fixed for the sake of others!

**Comment by ekwoka** [comment]
Interesting.

The alpineEffect in your workaround isn't doing anything, aside from maybe creating a memory leak.

x-html was also kind of not meant to be used lol certainly not to this degree.

I'll look at the change you identified and see if there's a way to get it working though.
